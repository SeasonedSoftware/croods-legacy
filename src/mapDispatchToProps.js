import { bindActionCreators } from 'redux'

import createActions from './createActions'

export default (dispatch, { options, ...props }) => ({
  actions: bindActionCreators(
    createActions({ ...options, ...props }),
    dispatch,
  ),
})
