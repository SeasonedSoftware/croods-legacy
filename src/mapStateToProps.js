import find from 'lodash/find'
import get from 'lodash/get'

export default (state, { name, parentId }) => {
  const statePart = get(state, name)
  const currentState = find(statePart, block => block.parentId === parentId)
  return get(currentState, 'state') || {}
}
