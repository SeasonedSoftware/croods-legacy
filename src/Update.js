import { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import customPropTypes from './customPropTypes'
import providerProps from './providerProps'
import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import resetIfChanged from './resetIfChanged'

class Update extends Component {
  constructor(props) {
    super(props)
    resetIfChanged({ props, prevProps: {}, name: 'updated' })
  }

  componentDidUpdate(prevProps) {
    resetIfChanged({ props: this.props, prevProps, name: 'updated' })
  }

  render() {
    const { render, id, attributes, actions } = this.props
    const update = () => actions.update({ id, ...attributes })

    return render(update, this.props)
  }
}

Update.propTypes = {
  /** Ex: 1234 or '1234' */
  id: customPropTypes.id.isRequired,
  /** Defines the deep properties used in the component. Ex: foo.bar */
  name: customPropTypes.name.isRequired,
  /** Properties to be updated */
  attributes: PropTypes.object.isRequired,
  /** A function returning a React Node. Ex: (update = (id, ...attributes), props) -> < JSX /> */
  render: PropTypes.func.isRequired,
  /**  Parent element's Id, for more specificity in changes.  */
  parentId: customPropTypes.id,
  /**  Ex: /foo/bar-_12?&= */
  path: customPropTypes.path,

  ...providerProps,
}

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Update),
)
