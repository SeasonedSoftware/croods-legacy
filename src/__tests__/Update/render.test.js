import React from 'react'
import { connect } from 'react-redux'
import renderer from 'react-test-renderer'

import Update from '../../Update'
import mapStateToProps from '../../mapStateToProps'
import mapDispatchToProps from '../../mapDispatchToProps'

jest.mock('../../withOptions', () => Component => Component)

jest.mock('react-redux', () => ({
  connect: jest.fn(() => Component => Component),
}))

const render = jest.fn((update, props) => {
  update && update()
  return (
    <div update={update} {...props}>
      Update
    </div>
  )
})

const actions = { update: jest.fn(), resetUpdated: jest.fn() }

beforeEach(() => {
  render.mockClear()
  actions.resetUpdated.mockClear()
})

it('renders correctly', () => {
  const props = { name: 'colors', attributes: {}, render, actions, id: 123 }
  const tree = renderer.create(<Update {...props} />).toJSON()

  expect(tree).toMatchSnapshot()
  expect(connect).toHaveBeenCalledWith(mapStateToProps, mapDispatchToProps)
  expect(actions.resetUpdated).not.toHaveBeenCalled()
  expect(actions.update).toHaveBeenCalledWith({ id: 123 })
})

describe('when updated', () => {
  it('resets updated', () => {
    const updated = { id: 123 }
    const props = {
      name: 'colors',
      attributes: {},
      render,
      actions,
      id: 123,
      updated,
    }
    const tree = renderer.create(<Update {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(connect).toHaveBeenCalledWith(mapStateToProps, mapDispatchToProps)
    expect(actions.resetUpdated).toHaveBeenCalledWith()
  })
})
