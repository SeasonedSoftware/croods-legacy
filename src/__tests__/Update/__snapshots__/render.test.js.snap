// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  actions={
    Object {
      "resetUpdated": [Function],
      "update": [Function],
    }
  }
  attributes={Object {}}
  id={123}
  name="colors"
  render={[Function]}
  update={[Function]}
>
  Update
</div>
`;

exports[`when updated resets updated 1`] = `
<div
  actions={
    Object {
      "resetUpdated": [Function],
      "update": [Function],
    }
  }
  attributes={Object {}}
  id={123}
  name="colors"
  render={[Function]}
  update={[Function]}
  updated={
    Object {
      "id": 123,
    }
  }
>
  Update
</div>
`;
